File: java-kotlin-project/app/src/main/java/example/java/basic/LambdaExample.java
package example.java.basic;


@FunctionalInterface
interface Calculator {
    int calculate(int a, int b);
}

@FunctionalInterface
interface Printer {
    void print(String msg);
}

public class LambdaExample {

    public void sample() {

        // 기본 람다 표현식
        Calculator add = (a, b) -> a + b;
        // Calculator add = Integer::sum;               // (메서드 참조형)
        Calculator multiply = (a, b) -> a * b;

        System.out.println(add.calculate(2, 4)); // 6
        System.out.println(multiply.calculate(2, 4)); // 8

        // N줄 람다 표현식
        Calculator complexCal = (a, b) -> {
            int result = a + b;
            result *= 2;
            return result;
        };

        System.out.println(complexCal.calculate(3, 4)); // 14

        // 매개변수가 없는 람다
        Runnable task = () -> System.out.println("Hello");

        // 매개변수가 하나인 람다
        Printer printer = msg -> System.out.println(msg);
        // Printer printer = System.out::println;       // (메서드 참조형)
    }

}

File: java-kotlin-project/app/src/main/java/example/kotlin/basic/LambdaExample.kt
package example.kotlin.basic

class LambdaExample {

    // 함수형 인터페이스 불필요 : 함수 타입 직접 사용 (함수 자체가 일급 객체다)
    fun main() {

        // 기본(람다) 표현식
        val add: (Int, Int) -> Int = { a, b -> a + b }
        val multiply = { a: Int, b: Int -> a * b }         // 타입 추론

        println(add(2, 4))  // 6
        println(multiply(2, 4)) // 8

        // N줄 람다 (마지막 표현식이 반환값)
        val complexCal = {
            a: Int, b: Int ->
            var result = a + b
            result *= 2
        }

        println(complexCal(2, 4)) // 12

        // 매개변수 없는 람다
        val task: () -> Unit = { println("Hello") }
        task()

        // 매개변수가 하나인 람다 (it 사용)
        val printer: (String) -> Unit = { println(it) }
        printer("hi!")

        // 명시적으로 이름 지정도 가능
        val printerWithName: (String) -> Unit = { msg -> println(msg) }
        printerWithName("hi~~")
    }
}